compiler/sem/sem.nim

```diff
@@ -694,6 +694,30 @@ proc semRealConstExpr(c: PContext, n: PNode): PNode =
   if result.kind != nkError:
     result = evalConstExpr(c, result)
 
+proc tryEvalStaticArgument(c: PContext, n: PNode): PNode =
+  ## Tries to evaluate an expression passed to a static parameter, while
+  ## overload resolution is still in progress. Returns nil if not successful.
+  var n = n
+  # `n` comes from sigmatch after a match and thus needs post-match fitting
+  if n.kind in {nkHiddenStdConv, nkHiddenSubConv, nkHiddenCallConv}:
+    # post-match fitting expects AST it can freely modify, which is guaranteed
+    # to be true for `n`, so copy the tree first
+    n = fitNodePostMatch(c, copyTree(n))
+  else:
+    n = copyNodeWithKids(n)
+
+  # prevent re-semming of the expression, which would throw away the
+  # types again:
+  n.flags.incl nfSem
+
+  result = tryConstExpr(c, n)
+  if result != nil:
+    let typ = newTypeS(tyStatic, c)
+    typ.sons = @[result.typ]
+    typ.n = result
+    result = copyNodeWithKids(n)
```

Comment:
Yep, we do, as `prepareNode` would run into infinite recursion otherwise. Consider the following situation:
1. `n` is passed to `tryConstExpr`
2. if `n` is already a constant expression and there's nothing to fold or evaluate, `tryConstExpr` returns `n` as-is
3. `result == n` is true in that case
4. a `tyStatic` with `PType.n` set to `result` (i.e., `n`) is created

If `result` would be set to `n` (without a copy), the `result.typ = typ` assignment below would result in a type where `typ.n.typ == typ`.

The shallow copy makes sure that `typ.n.typ != typ`, though I could change `result = copyNodeWithKids(n)` to:
```nim
result =
  if n != result: n
  else:           copyNodeWithKids(n)
```
in order to get around an unnecessary shallow copy, and to also emphasize why the copy is needed.

Reply:
Thanks for the details explanation, I forgot that `tryConstExpr` can act like an identity function in certain cases. I agree, the suggested code that avoids the copy is clearer in that regard.

Reply:
I've adjusted the code a bit -- it should be clearer now.